### PR TITLE
Basis Universal update

### PR DESCRIPTION
**Pull request by atteneder**
Updated the Basis Universal lib and fixed existing transcoding.

Note: I did not add support for additionally added formats, but it's a good base to start from.

The exact basisu source is a modified version waiting for PR review:
https://github.com/atteneder/basis_universal/tree/ktxfixes
commit 3d2ac2d23976e8ab4712f40781ae1e4eb723cf1b

### CODE REVIEW

**Comment by MarkCallow** [comment]
Did you use git subrepo to update the BasisU code? That is how the BasisU is incorporated in the KTX-Software repo.

**Comment by atteneder** [comment]
No I didn't, because I only knew/used GIT submodules before.
Now that I googled what subrepo is, I'll try to do a proper subrepo update later. Thanks for pointing it out to me.

**Comment by atteneder** [comment]
Current status: It's really a raw attempt to update with lots of bugs and crashes.
Just wanted to share early, in case others want to help out.

**Comment by MarkCallow** [comment]
Please rebase this PR on the ktx2 branch after the merge of PR #127. Sorry for giving you extra work.

**Comment by atteneder** [comment]
Desktop formats (compressed and uncompressed) mostly work on my machine. non power-of-two resolutions failed still. After some investigation I figured because the byte size calculations are off/wrong. I'm now trying to hack this in somehow, but in case someone has already figured a proper solution, I'm open to hearing it.

**Comment by MarkCallow** [comment]
PR #117 has one solution (for the same problem in `master`. However I prefer to use simple + and && instead of float conversions and ceil. 

The problem isn't directly related to NPOT. It occurs whenever a dimension is not a multiple of the block-size.

**Comment by MarkCallow** [comment]
PR #130 is now in progress to fix the image size calculation. Assuming no CI failures, it will  be merged shortly. + & & cannot be used because some of the formats, e.g. ASTC 5x5, have non power-of-two block sizes.

**Comment by MarkCallow** [comment]
@atteneder, I've pulled this branch into my workarea. After adding the new files to `libktx.gypi`, regenerating the Xcode projects and fixing some minor issues with the transcode enums I have it building successfully. Transcoding is not working properly. I suspect the way of initializing the transcoder has changed and I need to update `ktxTexture_TranscodeBasis` to whatever the new way is.

Did you write any new tests?

**Comment by atteneder** [comment]
Thanks for having a look.

No, I didn't write any tests. I'm a bit hesitant to get into setting up tests+gyp (I much rather continue working on my CMake branch).

Just pushed two smaller transcode fixes. In general, I have it working locally, but with some changes to the size/block count calculations. I'll let you know once I have something to show, but I'll look at the mentioned PRs first.

**Comment by atteneder** [comment]
fyi: I'm working on a test that transcodes a KTX2 as well as a basis file (same source content) via the default transcoder and compares the results/output. 7/19 tests fail atm, so some work left to do.

**Comment by MarkCallow** [comment]
> fyi: I'm working on a test that transcodes a KTX2 as well as a basis file (same source content) via the default transcoder and compares the results/output.

Sounds interesting. Does it, or do you have another too to, create the matching .ktx2 and .basis files as well? It's tedious to do by hand.

I do not have anything except transcode to uncompressed working in my copy of this branch. Transcode to BC3 has lots of artifacts. Transcode to ETC seems to give a texture with all alpha 0. Nothing shows up. I haven't had time for serious investigation. How are you testing transcoding?

**Comment by atteneder** [comment]
> Sounds interesting. Does it, or do you have another too to, create the matching .ktx2 and .basis files as well? It's tedious to do by hand.
right now I test the `color_grid.png` from testimages and created the basis file manually (via `basisu` command line tool)

> 
> I do not have anything except transcode to uncompressed working in my copy of this branch. Transcode to BC3 has lots of artifacts. Transcode to ETC seems to give a texture with all alpha 0. Nothing shows up. I haven't had time for serious investigation. How are you testing transcoding?

In my other branch (CMake work) I created C bindings I use for Unity3d. I started manually testing in the Unity editor (macOS; with some custom UI to load all kinds of formats) and made progress. After  that I built the Unity project for iOS and WebGL with partial success. Since the iteration time on those builds is way to high to do meaningful debuggin, I started writing the mentioned tests.

I'll try to wrap them up and make them available this week.

**Comment by MarkCallow** [comment]
I have transcode to ETC and ASTC working now. BC[13] are not working properly and I can't see the cause. Looking at the Basis examples, there doesn't appear to be any change in how to handle this format compared to before the transcoder updates. @atteneder have you tested with this format? This is what the transcoded result looks like:

Screen Shot 2019-10-15 at 22 37 44

**Comment by MarkCallow** [comment]
@atteneder which transcode target formats are you testing? Today I fixed the implementations of ETC2_RGBA and PVRTC1_R_RGBA in `ktxTexture_TranscodeBasis` both of which were badly broken. They now work. ETC1 and PVRTC1_4_RGB also work. I still have not found out why BC1 & BC3 do not work. Have you seen these work? I can't test any of the other formats due to lack of support in my hardware.

**Comment by MarkCallow** [comment]
I've created PR #131 with the changes I've made to this one.

File: tests/loadtests/glloadtests/shader-based/BasisuTest.cpp [hunk 1]
```diff
@@ -108,7 +108,7 @@ BasisuTest::BasisuTest(uint32_t width, uint32_t height,
     else if (features.etc1)
         tf = KTX_TF_ETC1;
     else if (features.pvrtc1)
-        tf = KTX_TF_PVRTC1_4_OPAQUE_ONLY;
+        tf = KTX_TF_BC7_M6_RGB;
```
Comment (MarkCallow) [inline tests/loadtests/glloadtests/shader-based/BasisuTest.cpp]: This is not correct. features.pvrtc1 will only ever be true if PVRTC1 support is detected and BC7 has nothing to do with PVRTC1. Further detection logic is needed in GL3LoadTestSample to detect BC7 whose result should be checked here. Please build on what I did in PR #131.

File: tests/loadtests/glloadtests/shader-based/DrawTexture.cpp [hunk 2]
```diff
@@ -111,7 +111,7 @@ DrawTexture::DrawTexture(uint32_t width, uint32_t height,
         else if (features.etc1)
             tf = KTX_TF_ETC1;
         else if (features.pvrtc1)
-            tf = KTX_TF_PVRTC1_4_OPAQUE_ONLY;
+            tf = KTX_TF_BC7_M6_RGB;
```
Comment (MarkCallow) [inline tests/loadtests/glloadtests/shader-based/DrawTexture.cpp]: Same comment as for BasisuTest.cpp.

**Comment by MarkCallow** [comment]
I fixed BC1 transcoding. The fix is in PR #131.

**Comment by atteneder** [comment]
I merged your recent work and now 16/18 tests succeed.
Great work!

**Comment by atteneder** [comment]
My transcode tests are currently in a dev branch of mine. I'm waiting for the basisu update to be merged before making a PR:
https://github.com/atteneder/KTX-Software/commit/00d5521950d74b81f0c9112cce44500d9283c11b

**Comment by MarkCallow** [comment]
I just landed PR #131 which incorporates this PR.

**Comment by atteneder** [comment]
Hi Mark,
See #133 
Thanks for your work on this one!

**Pull request closed without merging**